controllers/status_report.go

```diff
@@ -36,6 +36,13 @@ func (m *StatusReportMap) Set(id, reason, message string, hosts []string) {
 	}
 }
 
+func (m *StatusReportMap) ReadAll() map[string]StatusReport {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	return m.statuses
```

Comment:
Shouldn't be a copy? returning a reference, you are exposing the shared data and can be updated externally potentially running into race conditions errors.